test/integration/scheduler_perf/dra.go

```diff
@@ -364,3 +375,203 @@ claims:
+		}
+	}
+
+	return nil
+}
+
+func (op *updateDeviceConditionsOp) patchParams(w *workload) (realOp, error) {
+	op.Conditions = make([]metav1.Condition, 1)
+
+	var err error
+	if op.ConditionTypeParam != "" {
+		op.Conditions[0].Type, err = getParam[string](w.Params, op.ConditionTypeParam[1:])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if op.ConditionStatusParam != "" {
+		status, err := getParam[string](w.Params, op.ConditionStatusParam[1:])
+		if err != nil {
+			return nil, err
+		}
+		switch status {
+		case "True":
+			op.Conditions[0].Status = metav1.ConditionTrue
+		case "False":
+			op.Conditions[0].Status = metav1.ConditionFalse
+		case "Unknown":
+			op.Conditions[0].Status = metav1.ConditionUnknown
+		default:
+			return nil, fmt.Errorf("invalid condition status %q", status)
+		}
+	}
+
+	if op.DurationParam != "" {
+		durationStr, err := getParam[string](w.Params, op.DurationParam[1:])
+		if err != nil {
+			return nil, err
+		}
+		if op.Duration.Duration, err = time.ParseDuration(durationStr); err != nil {
+			return nil, fmt.Errorf("parsing duration parameter %s: %w", op.DurationParam, err)
+		}
+	}
+
+	if op.CountParam != "" {
+		op.Count, err = w.Params.get(op.CountParam[1:])
+		if err != nil {
+			return nil, fmt.Errorf("parsing CountParam %s: %w", op.CountParam, err)
+		} else if op.Count <= 0 {
+			return nil, fmt.Errorf("count must be greater than 0 if CountParam is set")
+		}
+	} else if op.Count <= 0 {
+		return nil, fmt.Errorf("count must be greater than 0 if CountParam is not set")
+	}
+
+	return op, op.isValid(false)
+}
+
+func (op *updateDeviceConditionsOp) collectsMetrics() bool {
+	return false
+}
+
+func (op *updateDeviceConditionsOp) requiredNamespaces() []string {
+	return []string{op.Namespace}
+}
+
+func (op *updateDeviceConditionsOp) run(tCtx ktesting.TContext) {
+	tCtx.Logf("updating device conditions in namespace %s with conditions %v", op.Namespace, op.Conditions)
+	var claimList *resourceapi.ResourceClaimList
+	var err error
+
+	ctx, cancel := context.WithTimeout(tCtx, op.Duration.Duration)
+	defer cancel()
+
+	err = wait.PollUntilContextTimeout(ctx, 1*time.Second, op.Duration.Duration, true, func(ctx context.Context) (bool, error) {
+		claimList, err = tCtx.Client().ResourceV1beta1().ResourceClaims(op.Namespace).List(tCtx, metav1.ListOptions{})
+		if err != nil {
+			return false, nil
+		}
+		if len(claimList.Items) >= op.Count {
+			return true, nil
+		}
+		tCtx.Logf("Waiting for at least %d claims in namespace %s, found %d", op.Count, op.Namespace, len(claimList.Items))
+		return false, nil
+	})
+
+	if err != nil {
+		tCtx.Fatalf("No claims found in namespace %s within %v: %v", op.Namespace, op.Duration.Duration, err)
+	}
+
+	for _, item := range claimList.Items {
+		claimName := item.GetName()
+		var claim *resourceapi.ResourceClaim
+
+		err = wait.PollUntilContextTimeout(ctx, 1*time.Second, op.Duration.Duration, true, func(ctx context.Context) (bool, error) {
```

Comment:
Do we plan to use these tests for performance? If so, how much will the result be influenced on how fast this code can update the conditions? I just want to understand what piece of code is actually being tested in that case.

Reply:
I'm not intending to use these tests for performance measurement.
However, it’s important that the code updates the conditions faster than the timeout value specified by BindingTimeoutSeconds. Otherwise, the test may unintentionally fail due to the Pod not being scheduled in time. So while performance isn’t the goal here, the timing does need to meet that threshold to ensure the test behaves as expected.

Reply:
The reason for my question is mostly around whether it would be better to handle integration tests as part of https://github.com/kubernetes/kubernetes/tree/master/test/integration/dra rather than here.

Reply:
"scheduler_perf" is good for tests scenarios which might become relevant for performance testing, because then those scenarios can be promoted to run in the periodic performance job.

To test correctness, `test/integration/dra` is now preferred. It's simpler to modify, in particular for one-of cases.

Compared to E2E tests, integration testing is preferred for binding conditions because integration tests are easier to run and debug. You can use your debugger to step through the kube-scheduler code! :smile: 

E2E tests are necessary when kubelet and actually running pods is required, so not really for binding conditiona. But a few E2E tests are needed for promotion to GA, because then we need conformance tests, which have to be in e2e.test.

Reply:
Just to confirm, based on this discussion, you would recommend implementing these tests in `test/integration/dra`, correct?

Regarding the discussion about test placement (scheduler_perf vs. test/integration/dra), I'm leaning towards moving these testcases to test/integration/dra since they primarily focus on correctness (and are unlikely to be used for performance testing in the future).

Reply:
Correct.